### PR TITLE
Sign entire cookie

### LINKED ISSUE #62: Feature: Sign entire cookie, not just value
I'd like signatures on cookies to include expiration date and path as well. The signature could still be inside the cookie value.

I would love to prepare a PR for this, but I'm unsure how to design this in a backwards compat way (or if that is even necessary).

### PR DESCRIPTION
**Pull request by untitaker**
Fix #62

### CODE REVIEW

File: src/jar.rs [hunk 1]
```diff
@@ -414,7 +414,7 @@ mod secure {
             let (text, signature) = match split_value(&cookie.value) {
                 Some(pair) => pair, None => return None
             };
-            let expected = dosign(key, text);
+            let expected = dosign(key, &cookie.to_string());
```
Comment (alexcrichton) [inline src/jar.rs]: Presumably this shouldn't work, right? We should be verifying the signature of `text` here but instead it's verifying a signature of a signature?

Comment (untitaker) [inline src/jar.rs]: You're right, I wonder why the tests pass though.

File: src/jar.rs [hunk 2]
```diff
@@ -540,17 +540,24 @@ mod test {
...
+            assert!($c.$secure().find("test").unwrap().value == "test");
+            let mut cookie = $c.find("test").unwrap();
+            cookie.expires = Some(cookie.expires.unwrap_or(0) + 1);
+            $c.add(cookie);
+            assert!($c.$secure().find("test").is_none());
         })
     }
 
-    #[cfg(features = "secure")]
+    #[cfg(feature = "secure")]
```
Comment (untitaker) [inline src/jar.rs]: This is so fucked up. Do you think I should file an issue against rustc for this?

File: src/jar.rs (hunk 1 above)
Comment (untitaker) [inline src/jar.rs]: Fixed!

**Comment by untitaker** [comment]
@alexcrichton ping

**Comment by alexcrichton** [comment]
Ah sorry haven't forgotten about this, just haven't had a chance to think through the backcompat story here yet

**Comment by untitaker** [comment]
@alexcrichton Bump. IMO it's not that big of a deal to invalidate all signed/encrypted cookies, as the browser doesn't guarantee either that those are stored. This is just session/cache expiration after all.

**Comment by alexcrichton** [comment]
Yeah that's true, let's just merge.

**Comment by untitaker** [comment]
Could you release a new version as well? Thanks!

**Comment by alexcrichton** [comment]
Sure, published 0.4.0

**Comment by alexcrichton** [comment]
Wait @untitaker this ended up breaking crates.io and now I'm confused. How does this work? Browsers don't send cookie parameters, just the values back. How is this supposed to be used?

**Comment by untitaker** [comment]
Ack, feel free to revert

**Comment by untitaker** [comment]
Sorry. I didn't test it locally, and assumed that the entire cookie including params would be sent by the client.

The idea (and my usecase) would be to securely expire a cookie without encoding the expiration date twice.

**Pull request merged**